src/typing/env_js.ml

```diff
@@ -327,7 +401,7 @@ let string_of_env cx ctx =
 let install_refinement cx x xtypes =
   if exists_in_scope x (peek_env ()) = None then
     let t = SMap.find_unsafe x xtypes in
-    init_env cx x (create_env_entry t t None)
+    init_env cx x (create_env_entry t t None) VarScope (* TODO: verify correct hoist behavior *)
```

Comment:
I'm pretty sure that the scope of the refinement depends on the type of entry corresponding to x. Furthermore, code that relies on `exists_in_scope`/`peek_env` instead of `exists_in_env`/`scope_kind` is generally suspect.

Unfortunately, we don't generally know the scope kind of the entry here, and I'm not even sure we can thread it through. The refinement logic is still a bit opaque to me, but I'll try to come up with some test cases to stress this code. Help here would be appreciated.